File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        l=sqrt(n)
        visited=[0]*(n+1)
        q=[(0,0)]
        while(q):
            s,l1=q.pop(0)
            if(s==n): return l1

            for i in range(1,int(l)+1):
                if(s+i**2 <=n):
                    if(visited[s+i**2]==0):
                        q.append((s+i**2 , l1+1))
                        visited[s+i**2]=1
    
### using dp ####
# class Solution1:
#     def numSquares(self, n: int) -> int:
        

File: 0279-perfect-squares/0279-perfect-squares.java
class Solution1 {
    public int numSquares(int n) {
        
        return solve(n);
    }
    private int solve(int n){
        
        if(n <= 1)  return n;   //base case
        int count = Integer.MAX_VALUE;
        for(int i = 1; i <= Math.sqrt(n); i++){
            int temp = 1 + solve(n - (i * i));
            count = Math.min(count, temp);
        }
        
        return count;
    }
}

class Solution {
    public int numSquares(int n) {
        
        int[] dp = new int[n + 1];
        Arrays.fill(dp, -1);
        return solve(n, dp);
    }
    private int solve(int n, int[] dp){
        
        if(n <= 1)  return n;   //base case
        if (dp[n] != -1)    return dp[n];
        int count = Integer.MAX_VALUE;
        for(int i = 1; i <= Math.sqrt(n); i++){
            int temp = 1 + solve(n - (i * i), dp);
            count = Math.min(count, temp);
        }
        
        return dp[n] = count;
    }
}